filters/block/matcher.go

```diff
@@ -0,0 +1,198 @@
+package block
+
+import (
+	"bytes"
+	"errors"
+	"io"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type maxBufferHandling int
+
+const (
+	readBufferSize uint64 = 8192
+)
+
+const (
+	maxBufferBestEffort maxBufferHandling = iota
+	maxBufferAbort
+)
+
+// matcher provides a reader that wraps an input reader, and blocks the request
+// if a pattern was found.
+//
+// It reads enough data until at least a complete match of the
+// pattern is met or the maxBufferSize is reached. When the pattern matches the entire
+// buffered input, the replaced content is returned to the caller when maxBufferSize is
+// reached. This also means that more replacements can happen than if we edited the
+// entire content in one piece, but this is necessary to be able to use the matcher for
+// input with unknown length.
+//
+// When the maxBufferHandling is set to maxBufferAbort, then the streaming is aborted
+// and the rest of the payload is dropped.
+//
+// To limit the number of repeated scans over the buffered data, the size of the
+// additional data read from the input grows exponentially with every iteration that
+// didn't result with any matched data blocked. If there was any matched data
+// the read size is reset to the initial value.
+//
+// When the input returns an error, e.g. EOF, the matcher finishes matching the buffered
+// data, blocks or return it to the caller.
+//
+// When the matcher is closed, it doesn't read anymore from the input or return any
+// buffered data. If the input implements io.Closer, closing the matcher closes the
+// input, too.
+//
+type matcher struct {
+
+	// init:
+	input             io.ReadCloser
+	matchList         []string
+	maxBufferSize     uint64
+	maxBufferHandling maxBufferHandling
+	readBuffer        []byte
+
+	// state:
+	ready   *bytes.Buffer
+	pending *bytes.Buffer
+
+	// final:
+	err    error
+	closed bool
+}
+
+var (
+	ErrMatcherBufferFull = errors.New("matcher buffer full")
+)
+
+func newMatcher(
+	input io.ReadCloser,
+	matchList []string,
+	maxBufferSize uint64,
+	mbh maxBufferHandling,
+) *matcher {
+
+	rsize := readBufferSize
+	if maxBufferSize < rsize {
+		rsize = maxBufferSize
+	}
+
+	return &matcher{
+		input:             input,
+		matchList:         matchList,
+		maxBufferSize:     maxBufferSize,
+		maxBufferHandling: mbh,
+		readBuffer:        make([]byte, rsize),
+		pending:           bytes.NewBuffer(nil),
+		ready:             bytes.NewBuffer(nil),
+	}
+}
+
+func (m *matcher) readNTimes(times int) (bool, error) {
+	var consumedInput bool
+	for i := 0; i < times; i++ {
+		n, err := m.input.Read(m.readBuffer)
+		m.pending.Write(m.readBuffer[:n])
+		if n > 0 {
+			consumedInput = true
+		}
+
+		if err != nil {
+			return consumedInput, err
+		}
+
+	}
+
+	return consumedInput, nil
+}
+
+func (m *matcher) match(b []byte) (int, error) {
+	var consumed int
+
+	for _, s := range m.matchList {
+		if bytes.Contains(b, []byte(s)) {
```

Comment:
We can prepare string s from matchList to be a []byte already, such that we do the conversion once.